### ISSUE TITLE
Class instance variables have no suggestions

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#1]
When declaring a variable like `$instance = new MyClass()` we should offer completion suggestions when typing `$instance->`

- [x] ~~**simple case** (`$instance = new MyClass()`)~~
- [x] ~~parameter type hinting (`function(MyClass $class) { }`)~~
- [x] ~~results of type hinted functions (`function something() : MyClass { }`)~~
- [x] **class properties** (`$this->prop = new MyClass()`) **`moved to` #68**
- [x] obscure edge cases (`(new MyClass)->myProperty()`) **`moved to` #68**

Branch: `feature/signatureHelp`
First pass merged in #67

**Comment by babe59** [comment]
Hello,
There is a chance that the bug is fixed quickly?
we must choose a PHP editor.
VSCode pleases me well.
By against the intellisense for the instantiated objects of my classes is indispensable

**Comment by nevadascout** [comment]
I'm working on it, but I'm afraid that I can't promise anything or give a potential completion date.

It could be done today, or it could take 6 months...

**Comment by babe59** [comment]
Ok Thanks
Good luck

**Comment by nevadascout** [comment]
First pass at this is now complete! It only supports local variables at the moment, but I'm working on getting properties working.

![](http://i.imgur.com/8mx7xDV.png)

**Comment by TheColorRed** [comment]
Not sure if you have taken this into account:

`(new MyClass)->myProperty();`

**Comment by TheColorRed** [comment]
Assuming this uses the namespaces to get the suggestions, this might need to be taken into account where two classes from different namespaces are used.

```php
using A\B\C;
using E\D\C as Z;

$instanceABC = new C;
$instanceEDC = new Z;
```

**Comment by nevadascout** [comment]
I hadn't realised you could do that. Seems like a terrible idea from a maintainability point of view, but I've added a note to the main issue.

It currently just uses the classname; if two classes have the same name then it will return suggestions for the first one it finds. I recognise this is an issue, proper namespace support is on the roadmap!

**Comment by TheColorRed** [comment]
Might also not want to forget method/function type hinting:

```php
public function myFunction(MyClass $class1, MyOther, $class2){
	$class1->func1();
	$class2->func1();
}
```

**Comment by nevadascout** [comment]
Added, thanks

**Comment by TheColorRed** [comment]
Thought of another one, when using the `use` with anonymous functions.

```php
$class1 = new Class1;
$class2 = new Class2;
public function myFunction(function() use ($class1, $class2){
    $class1->func1();
    $class2->func1();
});
```

**Comment by imperez** [comment]
@nevadascout Just curious on any update for this ticket/milestone? It looks like a great update and would love to have these features to use!

**Comment by nevadascout** [comment]
Hi @imperez, it's still in progress at the moment!

Once I get properties working then it will be rolled out as part of the 0.1.8 release - no idea when that will be though as I haven't got a huge amount of free time at the moment.

**Comment by nevadascout** [comment]
Resolved in #67
